oximeter/db/src/client.rs

```diff
@@ -642,34 +626,78 @@ impl DbWrite for Client {
             seen_timeseries.insert(key);
         }
 
-        // Insert the new schema into the database
-        //
-        // TODO-robustness There's still a race possible here. If two distinct clients receive new
-        // but conflicting schema, they will both try to insert those at some point into the schema
-        // tables. It's not clear how to handle this, since ClickHouse provides no transactions.
-        // This is unlikely to happen at this point, because the design is such that there will be
-        // a single `oximeter` instance, which has one client object, connected to a single
-        // ClickHouse server. But once we start replicating data, the window within which the race
-        // can occur is much larger, since it includes the time it takes ClickHouse to replicate
-        // data between nodes.
-        //
-        // NOTE: This is an issue even in the case where the schema don't conflict. Two clients may
-        // receive a sample with a new schema, and both would then try to insert that schema.
+        UnrolledSampleRows { new_schema, rows }
+    }
+
+    // Flush new schema to the database, or remove them from the cache on
+    // failure.
+    //
+    // This attempts to insert the provided schema into the timeseries schema
+    // table. If that fails, those schema are _also_ removed from the internal
+    // cache.
+    //
+    // TODO-robustness There's still a race possible here. If two distinct clients receive new
+    // but conflicting schema, they will both try to insert those at some point into the schema
+    // tables. It's not clear how to handle this, since ClickHouse provides no transactions.
+    // This is unlikely to happen at this point, because the design is such that there will be
+    // a single `oximeter` instance, which has one client object, connected to a single
+    // ClickHouse server. But once we start replicating data, the window within which the race
+    // can occur is much larger, since it includes the time it takes ClickHouse to replicate
+    // data between nodes.
+    //
+    // NOTE: This is an issue even in the case where the schema don't conflict. Two clients may
+    // receive a sample with a new schema, and both would then try to insert that schema.
+    async fn flush_new_schema_or_remove(
```

Comment:
To me `flush` connotes moving something from one place to another, so like if we're flushing to the DB, we're removing it from memory. Since I don't think that's the case here, I would call this `write` or `save` or something.

Reply:
`save` is fine too, yeah. I meant it in the same way _flush_ is used to describe a disk operation -- take all the things I asked the disk to do, which may have been buffered, and wait until they've been actually performed by the underlying storage medium.

Reply:
Renamed in 7699c2572